Only extend cluster database when reaching three member

ralubis opened the pull request:
Closes #6230

Signed-off-by: Rizwan Lubis <rizwan.lubis@gmail.com>
Signed-off-by: Jack Stenglein <jackstenglein@gmail.com>

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

freeekanayaka commented:
Nice job!

I think you got basically everything right except for a couple of things, one of which has turned out to be an existing bug that your work has uncovered :)

I can get your branch working (i.e. the second node is promoted only after the third node joins) by applying this diff on top of yours:

```diff
2 files changed, 7 insertions(+), 8 deletions(-)
lxd/api_cluster.go        | 6 ++++++
lxd/cluster/membership.go | 9 +--------

modified   lxd/api_cluster.go
@@ -375,6 +375,7 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) response.Response {
 		if err != nil {
 			return err
 		}
+		defer client.Close()
 
 		// If the ServerAddress field is set it means that we're using
 		// the new join API introduced with the 'clustering_join'
@@ -609,6 +610,11 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) response.Response {
 		version.UserAgentFeatures([]string{"cluster"})
 
 		logger.Infof("Making cluster rebalance call.")
+		client, err = cluster.Connect(req.ClusterAddress, d.endpoints.NetworkCert(), false)
+		if err != nil {
+			logger.Errorf("Failed to connect to cluster: %v", err)
+		}
+		defer client.Close()
 		err = clusterRebalance(client)
 		if err != nil {
 			logger.Errorf("Failed cluster rebalance call: %v", err)
modified   lxd/cluster/membership.go
@@ -656,16 +656,9 @@ func Promote(state *state.State, gateway *Gateway, nodes []db.RaftNode) error {
 
 	// Unlock regular access to our cluster database and add the database role.
 	err = state.Cluster.ExitExclusive(func(tx *db.ClusterTx) error {
-		logger.Info("Getting node list.")
-		nodeInfo, err := tx.Nodes()
-		if err != nil {
-			logger.Errorf("Failed to get node list: %v", err)
-			return errors.Wrapf(err, "Failed to get node list")
-		}
-		logger.Info("Test %d. Got node list: %+v", 0, nodeInfo)
 
 		logger.Info("Adding database role to this node")
-		err = tx.NodeAddRole(id, db.ClusterRoleDatabase)
+		err = tx.NodeAddRole(state.Cluster.GetNodeID(), db.ClusterRoleDatabase)
 		if err != nil {
 			logger.Errorf("Failed to add database role: %v", err)
 			return errors.Wrapf(err, "Failed to add database role for the node")
```

The ```client``` object that you were using was using the original joining node's TLS client certificate, so it was getting an "Unauthorized" error when trying to issue the rebalance request. If fixed it by creating a new ```client``` object that uses the cluster certificate as client certificate, which makes the rebalance request succeed since it now has the correct privileges.

I remove the extra logging since it was crashing (probably because you were using ```logger.Info``` instead of ```logger.Infof```, note the final ```f```).

Lastly, I fixed an existing bug, unrelated to your work: we were using the Raft ID when calling ```tx.NodeAddRole``` instead of the node database ID (the two IDs might differ). So you were getting the UNIQUE constraint error.

I've created a separate PR (#6486) to fix the existing issue. You'll want to rebase your branch on that one.

At this point, I think the only thing missing is fixing integration tests (the ones in ```clustering.sh```) since the new behavior of skipping the database role for the second node is most likely going to break them. You'll have to change the tests so they work with the new behavior.

jackstenglein commented:
@freeekanayaka @stgraber Hello, we think these commits should be sufficient. We cannot get the tests to run locally, but it looks like the Jenkins build passed and clustering.sh does not appear to check that the second node is a database node until the third has joined. Therefore, no tests in that file should be broken. Is this sufficient or should we try to get the tests running locally? Thank

stgraber commented:
Can you rebase this on master now that @freeekanayaka's branch has been merged?

Then we can do an actual jenkins run (only basic branch info was tested so far).

ralubis commented:
@stgraber Rebased to master.

stgraber commented:
Pretty sure that this will fail in clustering tests due to us expecting the database to replicate to the second node and us checking `cluster list` in a number of places. But lets see what happens on Jenkins.

stgraber commented:
jenkins: test this please

stgraber commented:
`membership_test.go` needs updating

ralubis commented:
jenkins: test this please

stgraber commented:
jenkins: test this please

freeekanayaka commented on lxd/api_cluster.go:
```diff
@@ -608,6 +608,15 @@ func clusterPutJoin(d *Daemon, req api.ClusterPut) response.Response {
+		client, err = cluster.Connect(req.ClusterAddress, d.endpoints.NetworkCert(), false)
+		if err != nil {
+			return err
+		}
```
Please add:

```
defer client.Close()
```

here just after checking for the error. Otherwise the underlying network connection will be leaked.

And while at it, please do the with the original ```client``` object, when it gets created earlier in the same function around line 377.

ralubis replied:
When we added `client.Close()` calls at lines 378 and 616, we got the errors `lxd/api_cluster.go:378:15: client.Close undefined (type lxd.InstanceServer has no field or method Close)
lxd/api_cluster.go:616:15: client.Close undefined (type lxd.InstanceServer has no field or method Close)
` when running `make`, so we did not originally include it. I am not sure if this is an error with my local environment, but I will add these changes to the branch.

freeekanayaka replied:
Oh sorry, I think I confused this ```client``` with another one. There is no ```Close()``` method indeed. Please scratch that, and revert the branch to your original version, otherwise it won't compile. Apologies for the confusion.

ralubis replied:
No worries, I reverted it just now.

freeekanayaka commented on lxd/cluster/membership.go:
```diff
@@ -198,8 +198,8 @@ func Accept(state *state.State, gateway *Gateway, name, address string, schema,
 	}
-
-	if len(nodes) < membershipMaxRaftNodes {
+	count, err := Count(state)
```
Please check the error, otherwise you end up ignoring it:

```go
if err != nil {
        return errors.Wrap(err, "Fetch cluster members count")
}
```

This should get rid of the static analysis failure reported by the test suite, look for:

```
/lxc-ci/build/tmp.blF58ZAwUA/go/src/github.com/lxc/lxd/lxd/cluster/membership.go:201:9: ineffectual assignment to err
```

in one of the logs, e.g. the [amd64/dir](https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/7666/arch=amd64,backend=dir,compiler=golang-1.13,restrict=lxd-priv/console) one.

freeekanayaka commented:
As mentioned before, you'll also need to fix the integration tests in ```test/suites/clustering.sh```, since lots of them assume that the 2nd node becomes a database member right away, not only after the 3rd is added.

For example you can see the ```clustering_address``` test fail here:

https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/7666/arch=amd64,backend=zfs,compiler=golang-1.13,restrict=lxd-priv/console#console-section-3

You should be able to reproduce that failure locally with something like:

```bash
cd test
sudo -E ./main.sh clustering_address
```

ralubis commented:
I changed the integration tests `test_clustering_recover()` and `test_clustering_address()` in `test/suites/clustering.sh` and squashed the merge into one commit for tests. All of the integration tests are passing locally, could we run it in Jenkins?

stgraber commented:
jenkins: test this please

stgraber commented:
Thanks!
